packages/actor-abstract-path/lib/ActorAbstractPath.ts

```diff
@@ -109,42 +109,71 @@ export abstract class ActorAbstractPath extends ActorQueryOperationTypedMediated
     });
   }
 
+  // Let the iterator `it` emit all bindings of size 2, with subjectStringVariable as value subjectVal
+  // and objectStringVariable as value all nodes reachable through predicate* beginning at objectVal
   public async ALPTwoVariables(subjectString: string, objectString: string, subjectVal: Term,
-    x: Term, predicate: Algebra.PropertyPathSymbol, context: ActionContext,
-    termHashes: {[id: string]: Term}, it: BufferedIterator<Bindings>, counter: any): Promise<void> {
-    const termString = termToString(subjectVal) + termToString(x);
-    if (termHashes[termString]) {
+    objectVal: Term, predicate: Algebra.PropertyPathSymbol, context: ActionContext,
+    termHashesGlobal: {[id: string]: Promise<Term[]>}, termHashesCurrentSubject: {[id: string]: boolean},
+    it: BufferedIterator<Bindings>, counter: any): Promise<void> {
+    const termString = termToString(objectVal);
+
+    // If this combination of subject and object already done, return nothing
+    if (termHashesCurrentSubject[termString]) {
       return;
     }
 
-    (<any> it)._push(Bindings({ [subjectString]: subjectVal, [objectString]: x }));
-    termHashes[termString] = x;
+    termHashesCurrentSubject[termString] = true;
+
+    (<any> it)._push(Bindings({ [subjectString]: subjectVal, [objectString]: objectVal }));
     counter.count++;
 
-    const thisVariable = this.generateVariable();
-    const vString = termToString(thisVariable);
-    const path = ActorAbstractPath.FACTORY.createPath(x, predicate, thisVariable);
-    const results = ActorQueryOperation.getSafeBindings(
-      await this.mediatorQueryOperation.mediate({ operation: path, context }),
-    );
-    results.bindingsStream.on('data', async bindings => {
-      const result = bindings.get(vString);
-      await this.ALPTwoVariables(
-        subjectString,
-        objectString,
-        subjectVal,
-        result,
-        predicate,
-        context,
-        termHashes,
-        it,
-        counter,
-      );
-    });
-    results.bindingsStream.on('end', () => {
-      if (--counter.count === 0) {
-        it.close();
+    // If every reachable node from object has already been calculated, use these for current subject too
+    if (termString in termHashesGlobal) {
+      const objects = await termHashesGlobal[termString];
+      for (const object of objects) {
+        if (!termHashesCurrentSubject[termToString(object)]) {
+          termHashesCurrentSubject[termToString(object)] = true;
+          (<any> it)._push(Bindings({ [subjectString]: subjectVal, [objectString]: object }));
+        }
       }
+    }
+
+    // Construct promise to calculate all reachable nodes from this object
```

Comment:
If I understand correctly, I think everything below the `if` statement should be in an `else` block? (Or the `if` statement should return out of the function). Otherwise I think you're going to keep overwriting the promises in `termHashesGlobal`. But besides that this seems to look good.

Reply:
You're right, there's missing a return;